sw/host/opentitansession/src/main.rs

```diff
@@ -96,6 +113,119 @@ fn parse_command_line(opts: Opts, mut args: ArgsOs) -> Result<Opts> {
     Ok(opts)
 }
 
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct SessionStartResult {
+    port: u16,
+}
+
+/// Spawn a child process, passing all the same arguments to the child, letting it instantiate a
+/// Transport based on the command line arguments, listen on a TCP socket, and run as a daemon
+/// process serving network requests.  Success of the child is verified by means of a
+/// `SessionStartResult` JSON message sent through the standard output pipe.
+fn start_session(run_file_fn: impl FnOnce(u16) -> PathBuf) -> Result<Box<dyn Serialize>> {
+    let mut args = args_os();
+
+    let mut child = Command::new(args.nth(0).unwrap()) // Same executable
+        .arg("--child") // Add argument to let the new process know it is the daemon child
+        .args(args) // Propagate all existing arguments: --interface, etc.
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    match serde_json::from_reader::<&mut ChildStdout, Result<SessionStartResult, String>>
+        (child.stdout.as_mut().unwrap())
+    {
+        Ok(Ok(result)) => {
+            // Create a pid file corresponding to the requested TCP port.
+            let path = run_file_fn(result.port);
+            File::create(path)?.write_all(format!("{}\n", child.id()).as_bytes())?;
+            Ok(Box::new(result))
+        }
+        Ok(Err(e)) => bail!(e),
+        Err(e) => bail!("Child process failed to start: {}", e),
+    }
+}
+
+// This method runs in the daemon child.  It will instantiate SessionHandler to bind to a
+// socket, then report the chosen port number to the parent process by means of a serialized
+// `SessionStartResult` sent through the stdout anonymous pipe, and finally enter an infnite
+// loop, processing connections on that socket
+fn session_child(listen_port: Option<u16>, backend_opts: &backend::BackendOpts) -> Result<()> {
+    let transport = backend::create(backend_opts, "null")?;
+    let mut session = SessionHandler::init(&transport, listen_port)?;
+    // Instantiation of Transport backen, and binding to a socket was successful, now go
+    // thought the process of making this process a daemon, disconeected from the
+    // terminal that was used to start it.
+
+    env::set_current_dir(&Path::new("/"))?;
+
+    // Close stdin and stderr, which would likely be connected to the tty (terminal)
```

Comment:
I want to keep the stderr of the child connected to the terminal, until it has completed the majority of its initialization.  That is, at least until it has constructed a Transport instance based on command line argument, and bound to a TCP port.  It could be that some error during initialization would trigger useful output on stderr, aside from whatever Err(_) it returns.  (Only the latter will be propagated to the parent process by means of the stdout pipe.)
stdin on the other hand, I do not need, so I may as well do `.stdin(Stdio::null())` from the start.

Reply:
Sounds good, as long as that's documented.

Reply:
Done